### DISCUSSION TITLE
How to set break point and debug mitmprxoy binary using pdb

### DISCUSSION THREAD

**Discussion by BabyTigerChaseDream** [mitmproxy/mitmproxy#6064] [Q&A]
Hello , I am trying to figure out the calling stack of mitmproxy, I want to set break point to mitmproxy's binary or mirmproxy.tools.main.py 
Can anyone suggest how can I do it with python pdb or any other debugging tools can used for it ?

@mhils  Is it possible to debug mitmproxy using pdb , if so , could you provide some guidance on it ? 
Thanks !

**Comment by ThomasBucaioni** [comment]
Hello,
What I did was to clone the repo, and add a few lines in `main.py`:
```
git clone https://github.com/mitmproxy/mitmproxy.git
```
as on this [page](https://github.com/mitmproxy/mitmproxy/blob/main/CONTRIBUTING.md#linux--macos).
Then in file `mitmproxy/tools/main.py`, add these lines in the `mitmweb` function:
```
def mitmweb(args=None) -> int | None:  # pragma: no cover
    from mitmproxy.tools import web
    
    #============
    #=== here ===
    #============
    import logging 
    import pdb; pdb.set_trace() 
    logging.debug('mitmproxy called from mitmweb')
    #============

    run(web.master.WebMaster, cmdline.mitmweb, args)
    return None
```
and run the virtual environment:
```
cd mitmproxy
python3 -m venv venv
venv/bin/pip install -e ".[dev]"
```
When I run 
```
$ mitmweb --version
> /home/user/github/mitmproxytests/mitmproxy/mitmproxy/tools/main.py(159)mitmweb()
-> logging.debug('mitmproxy called from mitmweb')
(Pdb)
```
Pdb shows the prompt. By pressing `c`, the execution resumes and the version appears:
```
Mitmproxy: 10.0.0.dev (+259, commit 0ace627)
Python:    3.11.3
OpenSSL:   OpenSSL 3.1.0 14 Mar 2023
Platform:  Linux-6.2.11-200.fc37.x86_64-x86_64-with-glibc2.36
```
Does it suit you? WebGui `mitmweb` is running with the breakpoint as I'm writing these lines and I can read myself in the proxy ;p
